policybot/cmd/server.go

```diff
@@ -115,3 +151,242 @@ func serverCmd() *cobra.Command {
 type dummyIoWriter struct{}
 
 func (dummyIoWriter) Write([]byte) (int, error) { return 0, nil }
+
+// Server represents a running bot instance.
+type server struct {
+	listener   net.Listener
+	shutdown   sync.WaitGroup
+	httpServer http.Server
+
+	clientID     string
+	clientSecret string
+	secretState  string
+
+	dashboard *dashboard.Dashboard
+}
+
+// Runs the server.
+//
+// If config comes from a container-based file, this will try to run the server, but if
+// problems occur (probably due to bad config), then the function returns with an error.
+//
+// If config comes from a repo-based file, this will also try to run the server, but if an error
+// occurs, it will refetch the config every minute and try again. And so in that case, this
+// function never returns.
+func runServer(base *config.Args) error {
+	for {
+		// copy the baseline config
+		c := *base
+		cfg := &c
+
+		// load the config file
+		if err := cfg.Fetch(); err != nil {
+			if cfg.StartupOptions.ConfigRepo != "" {
+				log.Errorf("Unable to load configuration file, waiting for 1 minute and then will try again: %v", err)
+				time.Sleep(time.Minute)
+				continue
+			} else {
+				return fmt.Errorf("unable to load configuration file: %v", err)
+			}
+		}
+
+		if err := runWithConfig(cfg); err != nil {
+			if cfg.StartupOptions.ConfigRepo != "" {
+				log.Errorf("Unable to initialize server likely due to bad config, waiting for 1 minute and then will try again: %v", err)
+				time.Sleep(time.Minute)
+			} else {
+				return fmt.Errorf("unable to initialize server: %v", err)
+			}
+		} else {
+			log.Infof("Configuration change detected, attempting to reload configuration")
+		}
+	}
+}
+
+func httpsRedirectHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		proto := req.Header.Get("x-forwarded-proto")
+		if proto == "http" || proto == "HTTP" {
+			http.Redirect(res, req, fmt.Sprintf("https://%s%s", req.Host, req.URL), http.StatusPermanentRedirect)
+			return
+		}
+
+		next.ServeHTTP(res, req)
+	})
+}
+
+func runWithConfig(a *config.Args) error {
+	log.Debugf("Starting with:\n%s", a)
+
+	creds, err := base64.StdEncoding.DecodeString(a.StartupOptions.GCPCredentials)
+	if err != nil {
+		return fmt.Errorf("unable to decode GCP credentials: %v", err)
+	}
+
+	ght := gh.NewThrottledClient(context.Background(), a.StartupOptions.GitHubToken)
```

Comment:
You're using a lot of context.Background(). I assume this is because these tasks are expected run until the process is torn down?

Reply:
I think so.

Is there a better pattern?

Reply:
If this is meant to be closed, you can create a sub-context that is tied to the lifecyle of the server. It looks like code already makes a lot of assumptions about the singleton aspect though, so it seems like a mute concern.